### DISCUSSION TITLE
Too-restrictive type for `getServerSideProps` preventing smart augmentation

### DISCUSSION THREAD

**Discussion by MattSchiller** [vercel/next.js#28285] [Help]
### What version of Next.js are you using?

11.0.1

### What version of Node.js are you using?

14.16.0

### What browser are you using?

Chrome (all)

### What operating system are you using?

OSX (any)

### How are you deploying your application?

next start

### Describe the Bug

It's a bug with the typings, which I think would be super easy to fix:

current:
```ts
export type GetServerSidePropsResult<P> =
  | { props: P }
  | { redirect: Redirect }
  | { notFound: true }
```

Proposed fix:
```ts
export type GetServerSidePropsResult<P> =
  | { props: P; redirect?: never; notFound?: never }
  | { redirect: Redirect; props?: never; notFound?: never }
  | { notFound: true; props?: never; redirect?: never }
```

I'm writing a helper method/wrapper for all of our pages' `getServerSideProps` methods to pass their page-specific stuff into to opt into app-wide augmentations to their props (adding things like the apollo cache or `req` data like cookies for SSR, etc).

But these methods return these mutually exclusive types, so I cannot easily do something like:
```
await passedGetServerSideProps(context).props
  ? doSomething()
  : doSomethingElse();
```

Because the `props` key doesn't exist on all flavors of the response.

### Expected Behavior

If the proposed `?: never`-typed definition was used, we'd retain identical type-checking around what a method is allowed to provide, but the key would crucially exist to allow typecheckers to not freak out about it.

### To Reproduce

Just try to access the `.props` or `redirect` or `notFound` keys on your `getServerSideProps` methods in a TypeScript-linted environment. You will not be able to.

**Comment by thebearingedge** [comment]
I have a similar wrapper scenario, but just ended up doing...

```ts
const result = await fancyGetServerSideProps(augmentedContext)
if ('props' in result) return doSomething(result)
if ('redirect' in result) return doSomethingElse(result)
return handleNotFound(result)
```

seems like a general enough solution

```ts
export const getServerSideProps = augmentContext(async fancyContext => {

})
```

**Reply by MattSchiller** [reply]
Yea, this is where we landed, too. It's definitely workable, just not as typesafe as I would've liked, since now `'props'` isn't tied to the `props` key on the interface in the same way, but I'll be satisfied with a 90% solution for sure =)

Thanks!
